xilem_core/src/sequence.rs

```diff
@@ -1,363 +1,651 @@
         }
+    }
+}
 
-        impl<'a, 'b> $elements_splice for $crate::VecSplice<'a, 'b, $pod> {
-            fn push(&mut self, element: $pod, _cx: &mut $cx) {
-                self.push(element);
-            }
+/// Views for ordered sequences of elements.
+///
+/// Generally, a container view will internally contain a `ViewSequence`.
+/// The child elements of the container will be updated by the `ViewSequence`.
+///
+/// This is implemented for:
+///  - Any single [`View`], where the view's element type
+///    is [compatible](SuperElement) with the sequence's element type.
+///    This is the root implementation, by which the sequence actually
+///    updates the relevant element.
+///  - An `Option` of a `ViewSequence` value.
+///    The elements of the inner sequence will be inserted into the
+///    sequence if the value is `Some`, and removed once the value is `None`.
+///  - A [`Vec`] of `ViewSequence` values.
+///    Note that this will have persistent allocation with size proportional
+///    to the *longest* `Vec` which is ever provided in the View tree, as this
+///    uses a generational indexing scheme.
+///  - Tuples of `ViewSequences` with up to 15 elements.
+///    These can be nested if an ad-hoc sequence of more than 15 sequences is needed.
+///
+pub trait ViewSequence<State, Action, Context: ViewPathTracker, Element: ViewElement, Marker>:
+    'static
+{
+    type SeqState;
+
+    /// Build the associated widgets into `elements` and initialize all states.
+    #[must_use]
+    fn seq_build(&self, ctx: &mut Context, elements: &mut AppendVec<Element>) -> Self::SeqState;
+
+    /// Update the associated widgets.
+    fn seq_rebuild(
+        &self,
+        prev: &Self,
+        seq_state: &mut Self::SeqState,
+        ctx: &mut Context,
+        elements: &mut impl ElementSplice<Element>,
+    );
+
+    /// Update the associated widgets.
+    fn seq_teardown(
+        &self,
+        seq_state: &mut Self::SeqState,
+        ctx: &mut Context,
+        elements: &mut impl ElementSplice<Element>,
+    );
+
+    /// Propagate a message.
+    ///
+    /// Handle a message, propagating to elements if needed. Here, `id_path` is a slice
+    /// of ids beginning at an element of this view_sequence.
+    fn seq_message(
+        &self,
+        seq_state: &mut Self::SeqState,
+        id_path: &[ViewId],
+        message: DynMessage,
+        app_state: &mut State,
+    ) -> MessageResult<Action>;
+}
 
-            fn mutate(&mut self, _cx: &mut $cx) -> &mut $pod
-            {
-                self.mutate()
-            }
+/// A temporary "splice" to add, update and delete in an (ordered) sequence of elements.
+/// It is mainly intended for view sequences.
+pub trait ElementSplice<Element: ViewElement> {
+    fn with_scratch<R>(&mut self, f: impl FnOnce(&mut AppendVec<Element>) -> R) -> R;
+    /// Insert a new element at the current index in the resulting collection.
+    fn push(&mut self, element: Element);
```

Comment:
Do we still need this, when there's `with_scratch` (which I'm not super happy with the naming, maybe it should be called `append` or this should be the new `push` instead)?
If we do keep this, it should either push at the end (this was the behavior before IIRC) or be called `insert` as the doc comment already implies IMO.

Reply:
I *think* it's needed to avoid additional e.g. allocations when they're not needed.

It's also a bit simpler to use. Although really, manual implementation of `ViewSequence` seem so unlikely that it wouldn't be unreasonable to remove this